[Skia] Add initial support for accelerated filters

GeorgesStavracas opened the pull request:
#### f8c0fe750d94b7db23d193c0b1f31858c2537620
<pre>
[Skia] Add initial support for accelerated filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=272535">https://bugs.webkit.org/show_bug.cgi?id=272535</a>

Reviewed by Carlos Garcia Campos.

This changeset enables accelerated filters on Skia builds.

Make WebCore::Page report FilterRenderingMode::Accelerated on Skia
builds if accelerated compositing is enabled.

Create Skia filters both in accelerated and software-based code paths,
since Skia is able to handle unaccelerated buffers transparently.

Add Skia-specific code to WebCore::FilterImage. The idea is that Skia
filters acquire a SkCanvas with WebCore::FilterImage::beginRecording(),
render the source image however they want, and finish the operation with
WebCore::FilterImage::finishRecording(). This generates a SkPicture
which is then used to paint on the resulting image buffer.

The memory cost is estimated using SkPicture. It&apos;s not an accurate
measure though.

Implement 3 filters as an exercise of usage of this new code:

 * SourceGraphicSkiaApplier, which merely copies the source image
 * FEColorMatrixSkiaApplier, for CSS and SVG filters like saturate(),
   greyscale(), contrast(), and others
 * FEGaussianBlurSkiaApplier, for the blur() filter

More filters will be added in follow up commits, as needed.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.cpp:
(WebCore::FEGaussianBlur::supportedFilterRenderingModes const):
(WebCore::FEGaussianBlur::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.h:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::memoryCost const):
(WebCore::FilterImage::imageBuffer):
* Source/WebCore/platform/graphics/filters/FilterImage.h:
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterRenderingModes const):
(WebCore::SourceGraphic::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp: Added.
(WebCore::FEColorMatrixSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.h: Added.
* Source/WebCore/platform/graphics/filters/skia/FEGaussianBlurSkiaApplier.cpp: Added.
(WebCore::FEGaussianBlurSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/FEGaussianBlurSkiaApplier.h: Added.
* Source/WebCore/platform/graphics/filters/skia/SourceGraphicSkiaApplier.cpp: Added.
(WebCore::SourceGraphicSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/SourceGraphicSkiaApplier.h: Added.
* Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp: Added.
(WebCore::FilterImage::beginRecording):
(WebCore::FilterImage::finishRecording):
(WebCore::FilterImage::memoryCostOfSkPicture const):
(WebCore::FilterImage::imageBufferFromSkPicture):

Canonical link: <a href="https://commits.webkit.org/277690@main">https://commits.webkit.org/277690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a39406f9aba1a85a6cca33ab0a34310369c2f5fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48269 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50957 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44334 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50574 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39430 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48851 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25175 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41702 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20580 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22655 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42882 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6325 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44613 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52861 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23316 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19668 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46767 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41884 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45691 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10655 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25386 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24304 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas commented on Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp:
```diff
@@ -0,0 +1,77 @@
+    if (!m_imageBuffer)
+        return nullptr;
+
+    // FIXME: is it safe to access m_imageBuffer->context().platformContext()?
```
This still stands - it's not clear to me if accessing `platformContext()` here is a mistake?

carlosgcampos replied:
Yes, it's safe

GeorgesStavracas replied:
Thanks

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp:
```diff
@@ -0,0 +1,77 @@
+    if (m_imageBuffer)
+        return m_imageBuffer.get();
+
+    m_imageBuffer = ImageBuffer::create<ImageBufferSkiaAcceleratedBackend>(m_absoluteImageRect.size(), 1, m_colorSpace, PixelFormat::BGRA8, RenderingPurpose::Unspecified, { });
```
Since we use the filters for non-accelerated graphics context too, I guess need to know somehow from here whether we should create an accelerated or unaccelerated image buffer.

GeorgesStavracas replied:
Done

carlosgcampos commented on Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp:
```diff
@@ -0,0 +1,109 @@
+        SkPaint paint;
+        paint.setColorFilter(SkColorFilters::Matrix(matrix.data()));
+
+        canvas->drawImage(nativeImage->platformImage().get(), 0, 0, SkSamplingOptions(), &paint);
```
There's a `SkCanvas::drawImage()` that receives a `sk_sp<SkImage>`, so you don't need the `get()`. You can also pass `{ }` instead of `SkSamplingOptions()`

GeorgesStavracas replied:
Done

TingPing commented on Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp:
```diff
@@ -0,0 +1,109 @@
+        paint.setColorFilter(SkColorFilters::Matrix(matrix.data()));
+
+        canvas->drawImage(nativeImage->platformImage(), 0, 0, { }, &paint);
+    }
```
Unnecessary scope?

Same comment for a few others.

TingPing replied:
Or was the intention `if (auto* canvas = result.beginRecording())`

GeorgesStavracas replied:
It's not necessary, yeah. I just wanted to visually keep the rendering code (i.e. everything that uses the `canvas` object) between `result.beingRecording()` and `result.finishRecording()`.

I thought `if (auto* canvas = result.beginRecording())` was a little deceiving because it will always be true. But now that I'm thinking, it'll also scope `canvas` into the if block. Might be better... thoughts?

TingPing replied:
> I thought if (auto* canvas = result.beginRecording()) was a little deceiving because it will always be true. 

Yes, only do that if its sometimes `nullptr`.

> But now that I'm thinking, it'll also scope canvas into the if block. Might be better... thoughts?

My opinion is drop the blocks all together. I don't think its a common pattern in webkit, unless its needed for a specific object in that scope like a lock.

GeorgesStavracas replied:
Done

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (m_imageBuffer)
+        return m_imageBuffer.get();
+
+    if (m_renderingMode == RenderingMode::Accelerated)
+        m_imageBuffer = ImageBuffer::create<ImageBufferSkiaAcceleratedBackend>(m_absoluteImageRect.size(), 1, m_colorSpace, PixelFormat::BGRA8, RenderingPurpose::Unspecified, { });
```
Isn't this always going to be accelerated? since we add it unconditionally, no? We need to know if the destination canvas is accelerated, not the Filter itself, no? Have you tried this in GTK with accelerated compositing disabled?

GeorgesStavracas replied:
Not anymore, and I can prove:
 * In Skia, if accelerated compositing is enabled
   * `WebCore::Page` will advertise `FilterRenderingMode::Software` and `FilterRenderingMode::Accelerated`
   * Accelerated filters (SourceGraphics, FEColorMatrix, and FEGaussianBlur) will advertise `FilterRenderingMode::Accelerated` as well
   * CSSFilter / SVGFilter will interscet both, and use `FilterEffect::createAcceleratedApplier()`, which creates an Skia effect applier if available
   * `RenderingMode::Accelerated` will be set, we try to use the Skia accelerated image buffer
 * If accelerated compositing is disabled:
   * `WebCore::Page` only advertises `FilterRenderingMode::Software`
   * * CSSFilter / SVGFilter will use `FilterEffect::createSoftwareApplier()`, which creates an Skia effect applier if available
   * `RenderingMode::Accelerated` will **not** be set, we use the Skia unaccelerated image buffer

Notice that there's `FilterRenderingMode`, and `RenderingMode`, and they're used in different ways.

The most notable difference in this latest push is that `WebCore::Page` now only advertises `FilterRenderingMode::Accelerated` if accelerated compositing is enabled (and even then, setting `WEBKIT_SKIA_ENABLE_CPU_RENDERING` disables accelerated images buffers).

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    return useCpuRendering.value();
+}
```
Instead fo reading this env var here again, you could just try to get the skia GL context, it will be nullptr when CPU rendering is enabled, but also when we fail to create the context for some reason, in which case we fall back to CPU rendering.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp:
```diff
@@ -0,0 +1,93 @@
+        return m_imageBuffer.get();
+
+    if (m_renderingMode == RenderingMode::Accelerated && !skiaUseCpuRendering())
+        m_imageBuffer = ImageBuffer::create<ImageBufferSkiaAcceleratedBackend>(m_absoluteImageRect.size(), 1, m_colorSpace, PixelFormat::BGRA8, RenderingPurpose::Unspecified, { });
```
Actually, we don't really need to explicitly check for cpu rendering here, because `ImageBufferSkiaAcceleratedBackend::create` returns nullptr when CPU rendering is enabled, because skia GL context is nullptr.

GeorgesStavracas replied:
You're right, it fails and falls back to the unaccelerated buffer path!

GeorgesStavracas replied:
Removed the code checking for the env var.

carlosgcampos reviewed (approved):
Looks good, let's just remove the cpu rendering check, I don't think it's needed.

GeorgesStavracas committed:
[Skia] Add initial support for accelerated filters
https://bugs.webkit.org/show_bug.cgi?id=272535

Reviewed by Carlos Garcia Campos.

This changeset enables accelerated filters on Skia builds.

Make WebCore::Page report FilterRenderingMode::Accelerated on Skia
builds if accelerated compositing is enabled.

Create Skia filters both in accelerated and software-based code paths,
since Skia is able to handle unaccelerated buffers transparently.

Add Skia-specific code to WebCore::FilterImage. The idea is that Skia
filters acquire a SkCanvas with WebCore::FilterImage::beginRecording(),
render the source image however they want, and finish the operation with
WebCore::FilterImage::finishRecording(). This generates a SkPicture
which is then used to paint on the resulting image buffer.

The memory cost is estimated using SkPicture. It's not an accurate
measure though.

Implement 3 filters as an exercise of usage of this new code:

 * SourceGraphicSkiaApplier, which merely copies the source image
 * FEColorMatrixSkiaApplier, for CSS and SVG filters like saturate(),
   greyscale(), contrast(), and others
 * FEGaussianBlurSkiaApplier, for the blur() filter

More filters will be added in follow up commits, as needed.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.cpp:
(WebCore::FEGaussianBlur::supportedFilterRenderingModes const):
(WebCore::FEGaussianBlur::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.h:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::memoryCost const):
(WebCore::FilterImage::imageBuffer):
* Source/WebCore/platform/graphics/filters/FilterImage.h:
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterRenderingModes const):
(WebCore::SourceGraphic::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp: Added.
(WebCore::FEColorMatrixSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.h: Added.
* Source/WebCore/platform/graphics/filters/skia/FEGaussianBlurSkiaApplier.cpp: Added.
(WebCore::FEGaussianBlurSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/FEGaussianBlurSkiaApplier.h: Added.
* Source/WebCore/platform/graphics/filters/skia/SourceGraphicSkiaApplier.cpp: Added.
(WebCore::SourceGraphicSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/SourceGraphicSkiaApplier.h: Added.
* Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp: Added.
(WebCore::FilterImage::beginRecording):
(WebCore::FilterImage::finishRecording):
(WebCore::FilterImage::memoryCostOfSkPicture const):
(WebCore::FilterImage::imageBufferFromSkPicture):

Canonical link: https://commits.webkit.org/277690@main
